neqo-transport/src/connection.rs

```diff
@@ -1261,38 +1268,22 @@ impl Connection {
                 Self::build_packet_header(path, *space, encoder, tx, &self.retry_info);
             let payload_start = builder.len();
 
-            // Work out how much space we have in the congestion window.
-            let limit = min(path.mtu(), cong_avail);
-            if builder.len() + tx.expansion() > limit {
-                // No space for a packet of this type in the congestion window.
+            // Work out if we have space left.
+            if builder.len() + tx.expansion() > profile.limit() {
+                // No space for a packet of this type.
                 encoder = builder.abort();
                 continue;
             }
-            let limit = limit - tx.expansion();
-
-            debug_assert!(!(pto && cc_limited));
 
             // Add frames to the packet.
-            let (tokens, ack_eliciting) = if *space >= min_pn_space {
-                let r = self.add_frames(&mut builder, *space, limit, cc_limited, now);
-                if builder.is_empty() {
-                    if pto {
-                        // Add a PING if there is a PTO and nothing to send.
-                        builder.encode_varint(Frame::Ping.get_type());
-                        (Vec::new(), true)
-                    } else {
-                        // Nothing to include in this packet.
-                        encoder = builder.abort();
-                        continue;
-                    }
-                } else {
-                    r
-                }
-            } else {
-                // A higher packet number space has a PTO; only add a PING.
-                builder.encode_varint(Frame::Ping.get_type());
-                (Vec::new(), true)
-            };
+            let limit = profile.limit() - tx.expansion();
+            let (tokens, ack_eliciting) =
+                self.add_frames(&mut builder, *space, limit, &profile, now);
+            if builder.is_empty() {
```

Comment:
It occurs to me that this is not a good check in the first place, because the builder will always include a header.  So I think we need to fix this test.

```suggestion
            if builder.len() == payload_start {
```